README.md
# Ember-time-tools

This Ember-cli addon provides date and time related tools for Ember applications.
The primary components consist of a date-picker, time-picker and a calendar.

## Demo

http://lozjackson.github.io/ember-time-tools/

## Installation

* `ember install ember-time-tools`

## Use

## DatePicker

Create a DatePickerComponent using the following example:

```
{{date-picker selectedDate=selectedDate select=(action "select")}}
```

You can use the InputDateComponent to create a html `input` element and a DatePickerComponent in one.

```
{{input-date value=date}}
```

## TimePicker

Create a TimePickerComponent using the following example:

```
{{time-picker selectedTime=selectedTime select=(action "select")}}
```

You can use the InputTimeComponent to create a html `input` element and a TimePickerComponent in one.

```
{{input-time value=time}}
```

## Calendar

You can create a CalendarMonthComponent using the following example:

```
{{calendar-month events=model}}
```

## Compatibility

Ember version ^2.0 is required for this addon to function correctly.  Ember v1.x is not supported.


CHANGELOG.md
# CHANGELOG

### v1.0.0 2018-04-15

* [#42](https://github.com/lozjackson/ember-time-tools/pull/42) Update Ember-cli to 2.18.2




### v0.8.1 2017-05-03

* [#39](https://github.com/lozjackson/ember-time-tools/pull/39) Rename container to avoid naming collisions.

* [#37](https://github.com/lozjackson/ember-time-tools/pull/37) Update Ember-cli to 2.12.1




### v0.8.0 2017-03-02

* [#36](https://github.com/lozjackson/ember-time-tools/pull/36) Update dependencies.

* [#34](https://github.com/lozjackson/ember-time-tools/pull/34) DatePickerComponent and TimePickerComponent should re-use selected value.




### v0.7.0 2017-01-27

* [#30](https://github.com/lozjackson/ember-time-tools/pull/30) Update TimePickerComponent header

* [#29](https://github.com/lozjackson/ember-time-tools/pull/29) [FEATURE] Add date-picker to CalendarMonthComponent.

* [#28](https://github.com/lozjackson/ember-time-tools/pull/28) Add Ember-cli-moment-shim dependency.

* [#27](https://github.com/lozjackson/ember-time-tools/pull/27) Update dependencies.

* [#26](https://github.com/lozjackson/ember-time-tools/pull/26) Update Ember-cli to 2.11.0




### v0.6.0 2017-01-02

* [#25](https://github.com/lozjackson/ember-time-tools/pull/25) Update `CalendarMonthComponent`.

* [#24](https://github.com/lozjackson/ember-time-tools/pull/24) Update Ember-ui-components dependency to ^0.11.0




### v0.5.0 2016-11-25

* [#23](https://github.com/lozjackson/ember-time-tools/pull/23) Update DatePickerComponent css styles.

* [#21](https://github.com/lozjackson/ember-time-tools/pull/21) [FEATURE] The DatePickerComponent now offers a list of months and years to select from.

* [#22](https://github.com/lozjackson/ember-time-tools/pull/22) Update Ember-ui-components dependency to ^0.10.0




### v0.4.0 2016-11-06

* [#20](https://github.com/lozjackson/ember-time-tools/pull/20) Add `scrollToSelectedTime` property to TimePickerComponent.

* [#18](https://github.com/lozjackson/ember-time-tools/pull/18) [FEATURE] Add SetPositionMixin

* [#18](https://github.com/lozjackson/ember-time-tools/pull/18) [FEATURE] Add `ember-wormhole` component to InputDateComponent and InputTimeComponent.

* [#16](https://github.com/lozjackson/ember-time-tools/pull/16) Add support for the placeholder attribute to InputDateComponent and InputTimeComponent.

* [#18](https://github.com/lozjackson/ember-time-tools/pull/18) Update Ember-ui-components dependency to ^0.9.2

* [#19](https://github.com/lozjackson/ember-time-tools/pull/19) Update to Ember-cli 2.8.0




### v0.3.0 2016-10-05

* [#14](https://github.com/lozjackson/ember-time-tools/pull/14) Add `format` property to InputDateComponent and InputTimeComponent.

* [#13](https://github.com/lozjackson/ember-time-tools/pull/13) InputDateComponent and InputTimeComponent now accept `disabled` attribute.

* [#12](https://github.com/lozjackson/ember-time-tools/pull/12) [BUGFIX] input components now allow empty values.




### v0.2.2 2016-09-12

* #5 Replace `Ember.Handlebars.SafeString` with `Ember.String.htmlSafe`.




### v0.2.1 2016-09-03

* [BUGFIX] Fix an issue that could cause the `nextMonth()` method to skip a month.

* Add Phantomjs dev-dependency.

* Configure Ember-try to include Ember versions ~2.0.0 and ~2.4.0 (LTS).

* Update to Ember-cli 2.7.0




### v0.2.0 2016-08-12

* [FEATURE] Add `timeInterval` property to the TimePickerComponent.

* [FEATURE] Add `displayFormat` property to the TimePickerComponent.

* Add  `timeInterval` and `pickerDisplayFormat` properties to the InputTimeComponent.

* Update the TimePickerComponent to refactor away from using a component for every time slot.

* Remove TimeSlotComponent




### v0.1.0 2016-06-26

* [FEATURE] Add `displayFormat` property to InputDateComponent and InputTimeComponent.

* [FEATURE] Update FormatDateHelper to allow passing in custom format strings.

* [FEATURE] When clicking outside the DatePickerComponent or TimePickerComponent the `close` action is triggered.




### v0.0.1 2016-06-26

Initial release


addon/utils/date.js
/**
  @module ember-time-tools
*/
import EmberObject, { computed } from '@ember/object';
import { typeOf } from '@ember/utils';

/**

  ```js
  let dateObject = DateObject.create();
  dateObject.setDate('2017/1/27');

  dateObject.get('month') // month = 0
  dateObject.set('month', 2) // month = 1
  dateObject.incrementProperty('month') // month = 2
  dateObject.decrementProperty('month') // month = 1
  dateObject.decrementProperty('month', 2) // month = 11, year = 2016
  ```

  @class DateObject
  @namespace Utils
*/
export default EmberObject.extend({

  /**
    javascript `Date()` object
    @property _date
    @type {Object}
    @private
  */
  _date: null,

  /**

    ```
    dateObject.set('date', '2016/7/22');
    dateObject.set('date', '2016-7-22');
    dateObject.set('date', new Date(2016, 6, 22));
    dateObject.set('date', 1469142000000);
    ```

    @property date
    @type {Object}
  */
  date: computed('_date', 'day', 'month', 'year', {
    get() {
      return this.get('_date');
    },
    set(key, value) {
      return this.setDate(value);
    }
  }),

  /**
    @property day
    @type {Integer}
  */
  day: computed('_date', {
    get() {
      return this.get('_date').getDate();
    },
    set(key, value) {
      this._modifyDate('Date', value);
      return this.get('_date').getDate();
    }
  }),

  /**
    @property month
    @type {Integer}
  */
  month: computed('_date', 'day', {
    get() {
      return this.get('_date').getMonth();
    },
    set(key, value) {
      this._modifyDate('Month', value);
      return this.get('_date').getMonth();
    }
  }),

  /**
    @property year
    @type {Integer}
  */
  year: computed('_date', 'month', 'day', {
    get() {
      return this.get('_date').getFullYear();
    },
    set(key, value) {
      this._modifyDate('FullYear', value);
      return this.get('_date').getFullYear();
    }
  }),

  /**
    @method init
    @private
  */
  init() {
    this._super(...arguments);
    this.set('_date', new Date());
  },

  /**
    ```
    dateObject.setDate('2016/7/22');
    dateObject.setDate('2016-7-22');
    dateObject.setDate(new Date(2016, 6, 22));
    dateObject.setDate(1469142000000);
    ```
    @method setDate
    @param {Object} date
    @return {Object} A javascript `Date` object
  */
  setDate(date) {
    if (typeOf(date) === 'string') {
      date = date.replace(/-/g, '/');
    } else if (typeOf(date) === 'instance') {
      date = new Date(date.get('year'), date.get('month'), date.get('date'));
    }
    return this.set('_date', new Date(date));
  },

  /**

    The `_modifyDate()` method takes two arguments.  A key and a value. The key specifies
    the part of the `Date()` object to set.  Can be either `Date`, `Month` or `FullYear`.

    ```js
    this._modifyDate('Month', 3); // set month to 3
    this._modifyDate('FullYear', 2017); // set year to 2017
    ```

    @method _modifyDate
    @param {String} key
    @param {Number} value
    @private
  */
  _modifyDate(key, value) {
    let date = this.get('_date');
    date[`set${ key }`](value);
    this.set('_date', date);
  }
});
